### DISCUSSION TITLE
HWP Supervisor Behavior

### DISCUSSION THREAD

**Discussion by jlashner** [simonsobs/socs#750] [Q&A]
Moving a few questions that Kyohei brought up in his HWP-supervisor update here.

What is a safe limit for the hwp_frequency (obtained from the PID controller) for allowing gripper action (@bbixler500)? Current threshold is set to 0.01 Hz, but I was unsure if the estimated hwp-freq was accurate at such low speeds. Do we know what the estimated freq does as the HWP comes to a stop, and at what point it goes from a trusted measurement to strictly zero?

What additional high-level functions should be added to the HWP Supervisor? Kyohei suggests some useful functions are:
- simple rotation function with constant drive voltage of hwp_pmx without pid.
- function to sweep hwp frequency from ? Hz to ? Hz over ? seconds.

**Comment by ykyohei** [comment]
> What is a safe limit for the hwp_frequency for allowing gripper action?

It's so true that neither hwp-pid nor hwp-encoder can provide an accurate hwp frequency at very low speeds.
I remembered that our plan was to measure the time since the hwp-pmx was turned off.

The HWP spindown time should definitely be less than 30 minutes so that we can safely grip after 30 minutes.

**Comment by ykyohei** [comment]
In the meeting on 5/12, as far as I understand, we concluded that if crossbar is working, we will shut down chwp gently by HWPsupervisor as follows. I think this is what we are developing & testing now.

- turn off pmx -> wait 30 mins -> grip -> turn off encoder led (turn off powerstrip, synnaccess)

When crossbar is not available, we will shut down more agressively by just turning off the pmx & powerstrip.

My questions are @jlashner 
1. Is "aggressive" shutdown possible by HWPsupervisor agent? Does it require crossbar?
2. Is it possible (necessary?) to implement a decision tree to choose whether we perform "gentle" shutdown and "aggressive" shutdown?

**Reply by jlashner** [reply]
Since brian moved this I realized I never responded to this, and since it still has yet to be implemented I may as well write it up.

1. I think the "aggressive shutdown" should not be implemented in the supervisor. The plan was (and I think still is) to have an agent running in the CHWP enclosure that is monitoring network status, and for it to manually disable ibootbar outlets when the network drops for too long.
2. I think this also answers the decision tree question... so supervisor will always try the gentle shutdown method. If that fails due to a networking issue, the agent running in the CHWP enclosure will be the fail-safe and will trigger independently.
